packages/pyright-internal/src/analyzer/sourceFile.ts

```diff
@@ -8,6 +8,7 @@
  */
 
 import { CancellationToken, CompletionItem, DocumentSymbol } from 'vscode-languageserver';
+import { TextDocument, TextDocumentContentChangeEvent } from 'vscode-languageserver-textdocument';
```

Comment:
Hmm, I was happy to have removed the language server notion of a TextDocument and TextDocumentContentChangeEvent from this file recently, but you've added them back. I don't think this is needed. The SourceFile can simply retain a string that holds its contents (and a second string that represents the "previous contents" if a modification is made while in edit mode).

Reply:
The problem with the 'previousContents' idea is you can't generate a very concise edit from it. It ends up being a full replace every time. This isn't so bad when the files are small, but sending full file contents back over LSP can be slow.

Reply:
Isn't there a way to generate a minimal set of text edits from a "before" and "after" file contents? I thought that's what Heejae was doing previously, as opposed to recording all of the individual edit actions.

Reply:
Heejae's code was doing the full thing:
https://github.com/microsoft/pyright/blob/3be44c3242aac5644931c243c5f7a0dc4cbe55cf/packages/pyright-internal/src/common/workspaceEditUtils.ts#L189

He suggested it could use the vscode differ, but I've used the vscode diff before when working on notebook concat stuff. It's very messy and doesn't always guess right (and ends up turning stuff into a full edit). 

The easier way is to just remember the edits.

Reply:
when you have nested edits, remembering edits won't work though if you were thinking sending those edits instead of sending whole documents.

if user change same file multiple time, you will need to adjust all spans since range would be wrong, and that's when there is no nested changes, if there are nested change, you need to merge those.

Reply:
I would suggest just sending the last full content replacing whole content rather than trying to be micro-optimize. if we really want to have smaller edits, we should just diff content and combine that data with tokens or parse nodes to find chunk we want to replace together.

Reply:
I'm not sure why the edits won't work? The same edits that were applied to change the file from one step to the next are returned.

Reply:
if you are thinking about sending each edits with different `applyEdit` so multiple changes including nested edits work, then, it will become multiple undo.

Reply:
So I think you're saying is that VS code doesn't handle the edits as one after another. The edits are actually all applied at once?

Reply:
edits and range are based on 1 version of document. not multiple version of document with changes.